File: data_structures/linked_list/typescript/DoublyLinkedListNode.ts
import SinglyLinkedListNode from '@ds/linked_list/typescript/SinglyLinkedListNode';
import { ILinkedListNode } from '@ds/linked_list/typescript/types';

export default class DoublyLinkedListNode<T> extends SinglyLinkedListNode<T> implements ILinkedListNode<T> {
    previous: DoublyLinkedListNode<T> | null = null;
    next: DoublyLinkedListNode<T> | null = null;

    constructor(
        value: T = undefined,
        key: number | string = null,
        next: DoublyLinkedListNode<T> | null = null,
        previous: DoublyLinkedListNode<T> | null = null,
    ) {
        super(value, key, next);

        this.next = next;
        this.previous = previous;
    }

    fromArray(values: T[]): this {
        let counter = 0;

        if (this.val === undefined) {
            this.val = values[0];
            counter = 1;
        }

        // eslint-disable-next-line @typescript-eslint/no-this-alias
        let current: DoublyLinkedListNode<T> | null = this;

        for (let i = counter; i < values.length; i++) {
            const node = new DoublyLinkedListNode(values[i]);

            node.previous = current;
            current.next = node;

            current = node;
        }

        return this;
    }

    reverse(): DoublyLinkedListNode<T> {
        if (!this.val) {
            return this;
        }

        // eslint-disable-next-line @typescript-eslint/no-this-alias
        let current: DoublyLinkedListNode<T> | null = this;
        let prev: DoublyLinkedListNode<T> | null = null;

        while (current) {
            const temp = current.next;
            current.next = current.previous;
            current.previous = temp;

            prev = current;
            current = current.previous;
        }

        return prev;
    }
}

File: data_structures/linked_list/python/doubly_linked_list_node.py
from typing import Union, List, TypeVar, Generic
from data_structures.linked_list.python.singly_linked_list_node import ListNode

T = TypeVar("T")


class DoublyLinkedListNode(ListNode, Generic[T]):
    next: Union[None, 'DoublyLinkedListNode'] = None
    previous: Union[None, 'DoublyLinkedListNode'] = None

    def __init__(self, val: T = None, key: int = None):
        super().__init__(val, key)

    def from_array(self, vals: List[T]) -> 'DoublyLinkedListNode[T]':
        """
        Create a singly linked list from array
        """
        counter = 0

        if len(vals) <= counter:
            pass

        if self.val is None:
            self.val = vals[0]
            counter += 1

        current = self

        for i in range(counter, len(vals)):
            node = DoublyLinkedListNode(vals[i])
            node.previous = current
            current.next = node
            current = node

        return self

    def reverse(self) -> 'DoublyLinkedListNode[T]':
        """
        Reverse doubly linked list
        """
        current_node = self
        prev_node = None

        while current_node:
            temp = current_node.next
            current_node.next = current_node.previous
            current_node.previous = temp

            prev_node = current_node
            current_node = current_node.previous

        return prev_node
